plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+                    done, pending = await asyncio.wait(pending, return_when=asyncio.FIRST_COMPLETED)
+                    for gathered_downloaders in done:
+                        results = gathered_downloaders.result()
+                        for download_result in results[:-1]:
+                            content = results[-1]
+                            for declarative_artifact in content.d_artifacts:
+                                if declarative_artifact.artifact._state.adding:
+                                    new_artifact = Artifact(
+                                        **download_result.artifact_attributes,
+                                        file=download_result.path
+                                    )
+                                    declarative_artifact.artifact = new_artifact
+                            pb.done = pb.done + len(content.d_artifacts)
+                            outstanding_downloads = outstanding_downloads - len(content.d_artifacts)
+                            await out_q.put(content)
+                else:
+                    if shutdown:
+                        break
+
+                if outstanding_downloads < self.max_concurrent_downloads:
+                    saturated = False
+        await out_q.put(None)
+
+
+async def artifact_saver(in_q, out_q):
+    """
+    Stages API stage that saves any unsaved :attr:`DeclarativeArtifact.artifact` objects.
+
+    This stage expects :class:`~pulpcore.plugin.stages.DeclarativeContent` units from `in_q` and
+    inspects their associated :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object stores one
+    :class:`~pulpcore.plugin.models.Artifact`.
+
+    Any unsaved :class:`~pulpcore.plugin.models.Artifact` objects are saved. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): The queue to receive
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        out_q (:class:`asyncio.Queue`): The queue to put
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+
+    Returns:
+        The artifact_saver stage as a coroutine to be included in a pipeline.
+    """
+    shutdown = False
+    declarative_content_list = []
+    while not shutdown:
+        try:
+            content = in_q.get_nowait()
+        except asyncio.QueueEmpty:
+            if not declarative_content_list:
+                content = await in_q.get()
+                declarative_content_list.append(content)
+                continue
+        else:
+            declarative_content_list.append(content)
+            continue
```

Comment:
This *batching* logic seems to rely on a momentary gap in the `in_q` to determine the batch size.  If the `in_q` is being fed faster than consumed here, won't that result in all of the content being in memory at once.  Conversely, what if the `in_q` is being fed in slow, small bursts.  Won't that result in really small batch sizes and poor performance?

Reply:
I agree but we should tune the stages later with data. I have a similar answer [here](https://github.com/pulp/pulp/pull/3559#discussion_r206942577).